### ISSUE TITLE
Early stopping criterion never reached?

### ISSUE THREAD

**Issue by nicolascedilnik** [MIC-DKFZ/nnUNet#585]
Hello. First of all, thanks for sharing this great piece of code.
I try training using my own dataset with the lowres model but it seems it never reaches the stopping criterion. By looking at this graph, I would say nothing improved since epoch 200.
![progress](https://user-images.githubusercontent.com/79838236/113383434-33dcb300-9384-11eb-9911-9ffa53593b23.png)

Is this expected? Is there something obviously when looking at this training graph? What would be the first parameter to tune to reach the stopping criterion of your great code? Thanks for your help!

EDIT: I found out that 1000 epochs were hardcoded for the nnUnetV2 trainer. It would be a nice improvement to add an easier way to switch to patience-based training than subclassing NetworkTrainer.

**Comment by BBQtime** [comment]
If I understood it correctly, the design of nnUNet deliberately to fits each fold of the training set thoroughly(1000 epochs). There may be two reasons, if I'm wrong, please adjust me: 

1. Guarantees the learning rate would decay to an optimal value for different tasks.
2. Train overly guarantees each fold/2D/3D step found the local minima at the mode of data, which helps the ensemble performance.

**Comment by FabianIsensee** [comment]
Hi,
there is no early stopping anymore. That was only part of the very first version of nnU-Net. The new (and better) nnU-Net uses hard coded 1000 epochs because that gives the best performance (on average across many datasets).
As @BBQtime rightfully pointed out, terminating the training early will not allow nnU-Net to go through the entire learning rate schedule. This is therefore not desired.
Best,
Fabian

**Comment by nicolascedilnik** [comment]
@FabianIsensee Thanks for the reply. I was thinking about doing:

```py
class MyFasterTrainer(nnunet.training.network_training.nnUNetTrainerV2.nnUNetTrainerV2):
    # Restore patience-based training completion instead of hardcoding 1000 epochs
    on_epoch_end = (
        nnunet.training.network_training.network_trainer.NetworkTrainer.on_epoch_end
    )
```

but I probably have to handle the learning rate schedule as well if I want to use a patience-based early stopping scheme. And you are saying this is a worst approach, so maybe I'll just stick to your 1000 hardcoded epochs.

**Comment by FabianIsensee** [comment]
Just set `self.max_num_epochs=XXX` in your init and that's it :-) Going back to early stopping is too complicated and will probably reduce your dice scores

**Comment by nicolascedilnik** [comment]
Great! Didn't think it could be so easy. Looking at the graph I have posted initially, do you agree that 1000 epochs is probably overkill for my task? Or would you go for 10,000 epochs if you had infinite computing power? ;-)

**Comment by FabianIsensee** [comment]
You would need to let it finish training. The learning rate goes down towards the end and this is typically reflected in the graph. This is the important part
At the end of the day you will only get definitive answers using trial and error. Given that one U-Net only trains for 2 days (1 day if you are on A100) I would say just let it run 1000 epochs. Whether longer training will help I don't know

**Comment by GregorKoehler** [comment]
Hi @nicolascedilnik,
from the looks of it, this issue has been resolved back in 2021. If you don't have any further questions here, we'll close the issue in 2 weeks.
